### ISSUE TITLE
Make File::Slurper dep optional?

### ISSUE THREAD

**Issue by ap** [benkasminbullock/gzip-libdeflate#2]
There are contexts where only the libdeflate binding is of interest whereas the convenience functions are irrelevant, in which case the extra dependency is actively undesirable.

**Comment by benkasminbullock** [comment]
Is it really such a problem? It's just a small pure-Perl module.

**Comment by ap** [comment]
Kinda. I understand why you don’t want to remove it, it makes things a bit ugly for shipping a small self-contained utility module. Users who want to use the convenience methods would have to depend on File::Slurper themselves, it’s kinda awful.

But on my end it’s also annoying to have this dead-weight transitive dependency if I want to use the module in a CPAN dist that is never going to use the convenience functions, like in a Plack middleware, you know? Then my users are installing and loading code that if nothing else uses it already is 100% useless to them.

I don’t know, is it an option maybe to split out a Gzip::Libdeflate::Raw or some such and make Gzip::Libdeflate a subclass of that? Kinda ungainly to have two separate dists I admit, but at least Gzip::Libdeflate itself could keep a fixed dep on File::Slurper and treat it as an implementation detail that downstream users don’t need to care about. So at least from a user perspective it would stay as neat as it already is.

**Comment by benkasminbullock** [comment]
This is the way the Perl language is usually done, there are dependencies on other CPAN modules in modules. File::Slurper seems to be a reasonable choice of module since it doesn't have any other dependencies, and it is well supported and tested. Rather than splitting this module into smaller pieces so that people don't have to download File::Slurper, couldn't you just bundle the required modules somehow?

**Comment by ap** [comment]
Uh, I’m [no stranger to Perl](https://metacpan.org/author/ARISTOTLE) exactly. I’m well aware how modules work, thank you.

If I create a Plack::Middleware::Whatever dist, are you suggesting I should bundle File::Slurper inside that? What good is that supposed to do? The couple kilobytes it takes to download the module and the couple seconds to install would be nice to avoid, sure, but they are not the issue. The real issue is that Gzip::Libdeflate loads File::Slurper unconditionally. Will bundling somehow defeat the `use File::Slurper` line when it’s useless?

The most I can think of why you would suggest that is that you think I am writing a Perl application. (It wouldn’t actually help anyway, even in that case) That is not my situation. My situation is that I would be writing a CPAN module. It would be using Gzip::Libdeflate internally, *without directly exposing it to the caller* – just the way your code is using File::Slurper internally.

Look at it from your own perspective. Would you be fine with it if `libdeflate` included a few small convenience functions for HTTP which require linking `libcurl` just because hey, why not?

The bottom line is that you are shipping a libedeflate binding… but actually the module is a libdeflate binding plus extra steps, and all I need (and want) is a binding for libdeflate itself.

**Comment by benkasminbullock** [comment]
Dependence on other modules is a standard practice in the Perl community. If this dependence causes you problems, the best I can offer you is comaintainer status, and you can split the module up into the pieces as you suggest. If that is acceptable to you, then let me know.

**Comment by ap** [comment]
That is entirely acceptable, absolutely.

**Comment by ap** [comment]
Sorry I wasn’t heard from again. I dropped the ball on this because the issue became moot for me: I was looking at this module as the basis for writing a replacement for Plack::Middleware::Deflater, but instead ended up switching from Gzip to Brotli, and in the process moving compression for dynamic responses out of our PSGI stack entirely, because Brotli has very nearly no story at all on CPAN.

Plus, another dependency I can neither remove nor pin at an older version has started loading File::Slurper as well, so while I’d still argue this would be a worthwhile thing to do in general, I wouldn’t be adding any benefit to our work codebase by doing it.

So my entire impetus here collapsed on multiple fronts and levels. Oh well.

**Comment by ap** [comment]
Sorry I wasn’t heard from again. I dropped the ball on this because the issue became moot for me:

I was looking at this module as the basis for writing a replacement for Plack::Middleware::Deflater, but instead ended up switching from Gzip to Brotli, and in the process moving compression for dynamic responses out of our PSGI stack entirely, because Brotli has very nearly no story at all on CPAN.

Plus, another dependency I can neither remove nor pin at an older version has started loading File::Slurper as well, so while I’d still argue this would be a worthwhile thing to do in general, I wouldn’t be adding any benefit to our work codebase by doing it.

So my entire impetus here collapsed on multiple fronts and levels. Oh well.
